src/transport_ofi.c

```diff
@@ -1252,17 +1251,25 @@ void shmem_transport_ctx_destroy(shmem_transport_ctx_t *ctx)
         shmem_free_list_destroy(ctx->bounce_buffers);
     }
 
-    ret = fi_close(&ctx->stx->fid);
-    OFI_CHECK_ERROR_MSG(ret, "STX context close failed (%s)\n", fi_strerror(errno));
+    if (ctx->stx) {
+        ret = fi_close(&ctx->stx->fid);
+        OFI_CHECK_ERROR_MSG(ret, "STX context close failed (%s)\n", fi_strerror(errno));
+    }
 
-    ret = fi_close(&ctx->put_cntr->fid);
-    OFI_CHECK_ERROR_MSG(ret, "Context put CNTR close failed (%s)\n", fi_strerror(errno));
+    if (ctx->put_cntr) {
+        ret = fi_close(&ctx->put_cntr->fid);
+        OFI_CHECK_ERROR_MSG(ret, "Context put CNTR close failed (%s)\n", fi_strerror(errno));
+    }
 
-    ret = fi_close(&ctx->get_cntr->fid);
-    OFI_CHECK_ERROR_MSG(ret, "Context get CNTR close failed (%s)\n", fi_strerror(errno));
+    if (ctx->get_cntr) {
+        ret = fi_close(&ctx->get_cntr->fid);
+        OFI_CHECK_ERROR_MSG(ret, "Context get CNTR close failed (%s)\n", fi_strerror(errno));
+    }
 
-    ret = fi_close(&ctx->cq->fid);
-    OFI_CHECK_ERROR_MSG(ret, "Context CQ close failed (%s)\n", fi_strerror(errno));
+    if (ctx->cq) {
```

Comment:
On OSX (sockets provider), this pointer is non-null after the `FI_EMFILE` error above, and trying to call `fi_close` on it results in a bad access seg. fault.  Seems like this approach might not work in general...

Reply:
I think @kseager called it!  We need to initialize the `shmem_transport_ctx_t` field pointers to NULL for this to work correctly.  My latest commit resolves the OSX issue.